pinot-core/src/main/java/com/linkedin/pinot/core/operator/BReusableFilteredDocIdSetOperator.java

```diff
@@ -62,36 +65,35 @@ public boolean open() {
   }
 
   @Override
-  public Block getNextBlock() {
+  public DocIdSetBlock getNextBlock() {
     // Handle limit 0 clause safely.
     // For limit 0, _docIdArray will be zero sized
-    if (_currentDoc == Constants.EOF) {
+    if (_currentDocId == Constants.EOF) {
       return null;
     }
     int[] docIdArray = DOC_ID_ARRAY.get();
-    if (!inited) {
-      inited = true;
-      _currentDoc = 0;
-      _currentBlock = _filterOperator.nextBlock();
-      _currentBlockDocIdIterator = _currentBlock.getBlockDocIdSet().iterator();
+    // Initialize filter block doc id set.
+    if (_filterBlockDocIdSet == null) {
+      _filterBlockDocIdSet = (FilterBlockDocIdSet) _filterOperator.nextBlock().getBlockDocIdSet();
+      _blockDocIdIterator = _filterBlockDocIdSet.iterator();
     }
     int pos = 0;
-    for (int i = 0; i < _maxSizeOfdocIdSet; i++) {
-      _currentDoc = _currentBlockDocIdIterator.next();
-      if (_currentDoc == Constants.EOF) {
+    for (int i = 0; i < _maxSizeOfDocIdSet; i++) {
+      _currentDocId = _blockDocIdIterator.next();
+      if (_currentDocId == Constants.EOF) {
         break;
       }
-      docIdArray[pos++] = _currentDoc;
+      docIdArray[pos++] = _currentDocId;
     }
-    DocIdSetBlock docIdSetBlock = null;
     if (pos > 0) {
-      docIdSetBlock = new DocIdSetBlock(docIdArray, pos);
+      return new DocIdSetBlock(docIdArray, pos, _filterBlockDocIdSet.getNumEntriesScannedInFilter());
```

Comment:
You are right, and we just use the result of the last DocIdSetBlock, which is the result after agg/group by/selection is finished.